lib/src/generate/watch_impl.dart

```diff
@@ -126,40 +127,53 @@ class WatchImpl {
       }
       assert(_nextBuildScheduled == false);
 
-      /// Remove any updates that were generated outputs or otherwise not
-      /// interesting.
-      var updatesToRemove = updatedInputs.keys.where(_shouldSkipInput).toList();
-      updatesToRemove.forEach(updatedInputs.remove);
-      if (updatedInputs.isEmpty && !force) {
+      /// Copy [updatedInputs] so that it doesn't get modified after this point.
+      /// Any further updates will be scheduled for the next build.
+      ///
+      /// Only copy the "interesting" outputs.
+      var updatedInputsCopy = <AssetId, ChangeType>{};
+      updatedInputs.forEach((input, changeType) {
+        if (_shouldSkipInput(input)) return;
+        updatedInputsCopy[input] = changeType;
+      });
+      updatedInputs.clear();
+      if (updatedInputsCopy.isEmpty && !force) {
         return;
       }
 
       _logger.info('Preparing for next build');
-      _logger.info('Clearing cache for invalidated assets');
-      void clearNodeAndDeps(AssetId id, ChangeType rootChangeType) {
+      Future clearNodeAndDeps(
+          AssetId id, AssetId primaryInput, ChangeType rootChangeType) async {
         var node = _assetGraph.get(id);
         if (node == null) return;
-        if (node is GeneratedAssetNode) {
-          node.needsUpdate = true;
-        }
         _assetCache.remove(id);
         for (var output in node.outputs) {
-          clearNodeAndDeps(output, rootChangeType);
+          await clearNodeAndDeps(output, primaryInput, rootChangeType);
```

Comment:
I thought there might be some race conditions here, but now that I think harder on it I don't think there will be, so I can update this.